packages/api-server/src/db/knex.ts

```diff
@@ -0,0 +1,315 @@
+      return {
+        type,
+        value: "undefined",
+      };
+
+    case DataType.STRING:
+      return {
+        type,
+        value: value as string,
+      };
+
+    case DataType.NUMBER:
+      return {
+        type,
+        value: value as number,
+      };
+
+    case DataType.BIGINT: {
+      const data = "0x" + (value as bigint).toString(16);
+      return {
+        type,
+        value: data,
+      };
+    }
+
+    case DataType.BOOLEAN: {
+      const data = (value as boolean).toString();
+      return {
+        type,
+        value: data,
+      };
+    }
+
+    case DataType.BUFFER: {
+      const data = "0x" + (value as Buffer).toString("hex");
+      return {
+        type,
+        value: data,
+      };
+    }
+
+    case DataType.DATE: {
+      const data = (value as Date).toString();
+      return {
+        type,
+        value: data,
+      };
+    }
+
+    case DataType.ARRAY: {
+      let data: SerializableData[] = (value as Array<DataType>).map((v) =>
+        serializeData(v)
+      );
+      return {
+        type,
+        value: data,
+      };
+    }
+
+    case DataType.OBJ: {
+      let data: any = {};
+      for (const k in value as any) {
+        let v = (value as any)[k];
+        data[k] = serializeData(v);
+      }
+      return {
+        type,
+        value: data,
+      };
+    }
+  }
+}
+
+function deserializeData(data: any): DataValue {
+  const { type, value } = data;
+
+  switch (type) {
+    case DataType.NULLABLE:
+      return undefined;
+
+    case DataType.STRING:
+      return value as string;
+
+    case DataType.NUMBER:
+      return +value;
+
+    case DataType.BIGINT: {
+      const data = BigInt(value);
+      return data;
+    }
+
+    case DataType.BOOLEAN: {
+      const data: boolean = JSON.parse(value);
+      return data;
+    }
+
+    case DataType.BUFFER: {
+      const data = Buffer.from(value.slice(2), "hex");
+      return data;
+    }
+
+    case DataType.DATE: {
+      const data = new Date(value);
+      return data;
+    }
+
+    case DataType.ARRAY: {
+      return (value as Array<any>).map((v) => deserializeData(v));
+    }
+
+    case DataType.OBJ: {
+      let data: any = {};
+      for (const k in value) {
+        let v: { type: DataType; value: any } = value[k];
+        data[k] = deserializeData(v);
+      }
+      return data as object;
+    }
+
+    default:
+      throw new Error("unsupported type: " + type);
+  }
+}
+
+interface Data {
+  type: DataType;
+  value: DataValue;
+}
+
+enum DataType {
+  NULLABLE,
+  STRING,
+  NUMBER,
+  BIGINT,
+  BOOLEAN,
+  BUFFER,
+  DATE,
+  ARRAY,
+  OBJ,
+}
+
+type DataValue =
+  | Nullable
+  | string
+  | number
+  | bigint
+  | boolean
+  | Buffer
+  | Date
+  | Array<DataValue>
+  | object;
+
+interface SerializableData {
+  type: DataType;
+  value: SerializableDataValue;
+}
+
+type SerializableDataValue = string | number | Array<SerializableData>;
+
+type Nullable = null | undefined;
```

Comment:
Name of nullable seems a little strange, it means a value can be null or not

Reply:
How about changing to more straightforward name `NullOrUndefined`